nomad/stream/ndjson.go

```diff
@@ -0,0 +1,105 @@
+package stream
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+var (
+	// NDJsonHeartbeat is the NDJson to send as a heartbeat
+	// Avoids creating many heartbeat instances
+	NDJsonHeartbeat = &structs.NDJson{Data: []byte("{}\n")}
+)
+
+// NDJsonStream is used to send new line delimited JSON and heartbeats
+// to a destination (out channel)
+type NDJsonStream struct {
+	// mu locks during send and protects running
+	mu sync.Mutex
+
+	out chan<- *structs.NDJson
+
+	// heartbeat is the interval to send heartbeat messages to keep a connection
+	// open.
+	heartbeat *time.Ticker
+
+	publishCh chan structs.NDJson
+	exitCh    chan struct{}
+
+	running bool
+}
+
+// NewNNewNDJsonStream creates a new NDJson stream that will output NDJson structs
+// to the passed output channel
+func NewNDJsonStream(out chan<- *structs.NDJson, heartbeat time.Duration) *NDJsonStream {
```

Comment:
I'd consider having `out` channel be owned by this struct: it can be an internally created channel, exposed in a method.  Simplifies cllients of the class a bit.

Reply:
I liek that too, changing